elements/lisk-tree/src/merkle_tree.ts

```diff
@@ -83,12 +176,159 @@ export class MerkleTree {
 	}
 
 	public clear(): void {
-		this._dataLength = 0;
-		this._data = {};
+		this._width = 0;
+		this._hashToBuffer = {};
 		this._root = EMPTY_HASH;
 	}
 
+	public getData(): object[] {
+		return Object.keys(this._hashToBuffer).map(key =>
+			this.getNode(Buffer.from(key, 'binary')),
+		);
+	}
+
+	public toString(): string {
+		if (this._width === 0) {
+			return this.root.toString('hex');
+		}
+		return this._printNode(this.root);
+	}
+
 	private _getHeight(): number {
-		return Math.ceil(Math.log2(this._dataLength)) + 1;
+		return Math.ceil(Math.log2(this._width)) + 1;
+	}
+
+	private _generateLeaf(value: Buffer): NodeData {
+		const leafValue = Buffer.concat(
+			[LEAF_PREFIX, value],
+			LEAF_PREFIX.length + value.length,
+		);
+		const leafHash = hash(leafValue);
+		this._hashToBuffer[leafHash.toString('binary')] = leafValue;
+		this._width += 1;
+
+		return {
+			value: leafValue,
+			hash: hash(leafValue),
+		};
+	}
+
+	private _generateNode(
```

Comment:
Why not `generateBranch`?

Reply:
I thought of naming it `generateBranch`, but it also generates possible root node. It seems semantically awkward to call a root node a branch.

Reply:
It can?

Reply:
`generateLeaf` will only be called during initial `build` or otherwise when `append` is called.